core/src/main/java/com/graphhopper/routing/weighting/custom/SpeedAndAccessProvider.java

```diff
@@ -15,19 +15,13 @@
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  */
+
 package com.graphhopper.routing.weighting.custom;
 
-import com.graphhopper.routing.ev.EncodedValue;
 import com.graphhopper.util.EdgeIteratorState;
 
-/**
- * Implementations of this internal interface represent the conversion rules we use to map {@link EncodedValue}s to
- * numerical values when determining the different parameters for {@link CustomWeighting}.
- */
-interface EdgeToValueEntry {
+interface SpeedAndAccessProvider {
```

Comment:
Should we rename to SpeedAndPriorityProvider or even something else :) ?

Reply:
Oh yes, no idea why its called SpeedAndAccess... 

> or even something else 

Do you have an idea?

Reply:
We can delete it without much pain it seems :)

Reply:
Ok good :) So ExpressionBuilder now knows about CustomWeighting.Parameters but CustomWeighting is still decoupled from CustomWeightingHelper and janino stuff :+1:. We can probably now move the static CustomWeighting#create into ExpressionBuilder?

And can we rename ExpressionBuilder?

Reply:
> We can probably now move the static CustomWeighting#create into ExpressionBuilder?

Sure.

> And can we rename ExpressionBuilder?

Do you have a suggestion? SpeedAndPriorityProvider :D ?

Reply:
Haha :D

On this level it would be the CustomWeightingParameterBuilder, but that is not very helpful. This class seems to be the one that takes a custom model and gives a meaning to it so I like something like `CustomModelParser` or `JaninoInterpreter` or something. I do not like `ExpressionBuilder` because 'expression' seems very unspecific and even more it is not building expressions :)

Reply:
> We can probably now move the static CustomWeighting#create into ExpressionBuilder?

Hmmh, it seems this method is used in DefaultWeightingFactory and so the ExpressionBuilder class would need to be public which would be suboptimal.

Reply:
Would it? The weighting factory already knows there is a custom model so why should it not know how to create a weighting from it (thats its job really). I rather think CustomWeighting should not know how to build itself so actually CustomWeighting should not know about ExpressionBuilder. Maybe its just in the wrong package then?

Reply:
I like CustomModelParser :) 

> I rather think CustomWeighting should not know about janino etc. so actually CustomWeighting should not know about ExpressionBuilder.

Ok, yes :) 

> Btw did you try to make CustomWeightingHelper package level or at least protected? That one really seems to be an implementation detail.

Yes, unfortunately it wasn't possible due to janino magic :/

Reply:
> Btw did you try to make CustomWeightingHelper package level or at least protected?

When package protected it says something like:

```
java.lang.IllegalAccessError: class com.graphhopper.routing.weighting.custom.JaninoCustomWeightingHelperSubclass2 cannot access its superclass com.graphhopper.routing.weighting.custom.CustomWeightingHelper (com.graphhopper.routing.weighting.custom.JaninoCustomWeightingHelperSubclass2 is in unnamed module of loader org.codehaus.commons.compiler.util.reflect.ByteArrayClassLoader @5c86dbc5; com.graphhopper.routing.weighting.custom.CustomWeightingHelper is in unnamed module of loader 'app')
```

and classes cannot be protected it seems.

Reply:
Haved renamed ExpressionBuilder to CustomModelParser, made it public and moved the `create` method there. Is this better?

Reply:
Yes I think so :+1:

Just two minor things:

1) You probably tried this, but does the __internal_cache method affect json/yaml serialization? Or is this the reason for the __ prefix?

2) Maybe rename CustomModelParser#create to CustomModelParser#createWeighting?

Reply:
> You probably tried this, but does the __internal_cache method affect json/yaml serialization? 

It is not a setter or getter so I do not think it is affected and the tests are still passing. Or are you worried that the "internal_cache" attribute could be set via a request?

> Or is this the reason for the __ prefix?

The reason is more that it is not suggested in an auto completion and looking ugly so it should force a "responsible" developer using it to look into the javadocs :) ?

Hmmh, or should we move CustomModel to the `com.graphhopper.routing.weighting.custom` package and make it package protected?

Reply:
Jackson either uses a naming convention and assumes there is a field when ever there are setters/getters, or it looks at the fields directly (I don't know what the default is).

>  Or are you worried that the "internal_cache" attribute could be set via a deserialization i.e. via a request?

No I was not, as long as Jackson uses method access and the getters/setters do not follow the naming convention it should not be a problem.

> The reason is more that it is not suggested in an auto completion and looking ugly so it should force a "responsible" developer using it to look into the javadocs :) ?

Ok yes makes sense to me.

> Hmmh, or should we move CustomModel to the com.graphhopper.routing.weighting.custom package and make this method package protected?

Which method?

Reply:
> Which method?

We could hide both: `CustomModel.__internal_cache` and `__shouldBeCached`. And CustomProfile and CustomModelParser are already in that package.

Reply:
I actually like this - see latest commit.